source/dagre.js

```diff
@@ -1349,6 +1349,110 @@ dagre.layout = (graph, layout) => {
+                            continue;
+                        }
+                        exchange(layer1, node1u, node2u);
+                        exchange(layer2, node1, node2);
+                        exchange(layer3, node1d, node2d);
+                        ++k;
+                    }
+                }
+                for (let j = 0; j < layer2.length - 1; ++j) {
+                    const node0 = g.nodes.get(layer2[j]);
+                    if (!node0.in || !node0.out || node0.in.length !== 1 || node0.out.length !== 1) {
+                        continue;
+                    }
+                    const node1 = g.nodes.get(layer2[j + 1]);
+                    if (!node1.in || !node1.out || node1.in.length !== 1 || node1.out.length !== 1) {
+                        continue;
+                    }
+                    const node0u = node0.in[0].vNode;
+                    const node1u = node1.in[0].vNode;
+                    const node2 = node0u.in[0].vNode;
+                    const node3 = node1u.in[0].vNode;
+                    let idx0 = layer0.indexOf(node2.v);
+                    let idx1 = layer0.indexOf(node3.v);
+                    if (idx1 + 1 !== idx0) {
+                        continue;
+                    }
+                    const node0d = node0.out[0].wNode;
+                    const node1d = node1.out[0].wNode;
+                    idx0 = layer3.indexOf(node0d.v);
+                    idx1 = layer3.indexOf(node1d.v);
+                    if (idx0 + 1 !== idx1) {
+                        continue;
+                    }
+                    if (node0d.out[0].wNode !== node1d.out[0].wNode) {
+                        continue;
+                    }
+                    exchange(layer1, node0u, node1u);
+                    exchange(layer2, node0, node1);
+                    exchange(layer3, node0d, node1d);
+                    ++j;
+                }
+            }
+            // mitigate crossings of input connection lines
+            for (let i = 0; i < best.length - 2; i += 2) {
+                const layer0 = best[i];
+                const layer1 = best[i + 1];
+                const layer2 = best[i + 2];
+                for (let j = 0; j < layer2.length; ++j) {
+                    const node0 = g.nodes.get(layer2[j]);
+                    if (!node0.in || node0.in.length < 2) {
+                        continue;
+                    }
+                    for (let k = 0; k < node0.in.length - 1; ++k) {
+                        const node1d = node0.in[k].vNode;
+                        const node2d = node0.in[k + 1].vNode;
+                        const node1 = node1d.in[0].vNode;
+                        const node2 = node2d.in[0].vNode;
+                        if ((layer1.indexOf(node1d.v) < layer1.indexOf(node2d.v)) ^ (layer0.indexOf(node1.v) < layer0.indexOf(node2.v))) {
+                            exchange(layer1, node1d, node2d);
+                        }
+                    }
```

Comment:
Comparing adjacent input connection nodes isn't enough. Sorting input connection nodes would be needed.

Reply:
It's probablly not so easy to compare the order of input nodes among multiple layers. So I'll resolve this thread.